### PR TITLE
fix several audit issues

### PR DESCRIPTION
**Pull request by chaosma**
* [done] Inconsistent restriction on voice credit upper bound: remove the unnecessary check
* [done] Data are not fully verified during state update: modified and fixed with new logics
* [done] change some parameters in TopupCredit.sol and MACI.sol to constant variable to save gas
* [done] Token for top-up is a free resource: add onlyOwner modifier on airdrop and airdropTo
* [done] MessageQueue in PollFactory is uninitialized: add enqueue of NOTHING_UP_MY_SLEEVE number

### CODE REVIEW

File: circuits/circom/processMessages.circom [hunk 1]
```diff
@@ -91,12 +91,6 @@ template ProcessMessages(
     // The state root before it is processed
     signal input currentStateRoot;
 
-    // topup signals
```
Comment (Mikerah) [inline circuits/circom/processMessages.circom:94]: The associated tests should be updated to ensure that this change hasn't affected any invariants. The suggestions for this issue have been applied but in order to ensure that the behavior we descrcibe no longer is a problem, a test case should be written

File: contracts/contracts/MACI.sol [hunk 2]
```diff
@@ -219,15 +214,6 @@ contract MACI is IMACI, DomainObjs, Params, SnarkCommon, Ownable {
             _initialVoiceCreditProxyData
         );
 
-        // The limit on voice credits is 2 ^ 32 which is hardcoded into the
-        // MessageValidator circuit, specifically at check that there are
-        // sufficient voice credits (using GreaterEqThan(32)).
-        // TODO: perhaps increase this to 2 ^ 50 = 1125899906842624?
-        require(
```
Comment (Mikerah) [inline contracts/contracts/MACI.sol:226]: Our suggestion here was to be consistent across the codebase. However, considering the gas constraints, are you sure you want to unrestrict the number of voice credits? There also a valid argument that such enforcement can be done within the front-end and removing the check within the contract actually saves gas. Is that the reasoning for applying this fix?

File: contracts/contracts/TopupCredit.sol [hunk 3]
```diff
@@ -15,12 +15,12 @@ contract TopupCredit is ERC20, Ownable {
         return _decimals;
     }
 
-    function airdropTo(address account, uint256 amount) public {
+    function airdropTo(address account, uint256 amount) public onlyOwner {
```
Comment (Mikerah) [inline contracts/contracts/TopupCredit.sol:18]: LGTM

File: contracts/contracts/TopupCredit.sol [hunk 4]
```diff
@@ -15,12 +15,12 @@ contract TopupCredit is ERC20, Ownable {
...
         require(amount < MAXIMUM_AIRDROP_AMOUNT);
         _mint(account, amount);
     }
 
-    function airdrop(uint256 amount) public {
+    function airdrop(uint256 amount) public onlyOwner {
```
Comment (Mikerah) [inline contracts/contracts/TopupCredit.sol:23]: LGTM

File: contracts/contracts/Poll.sol [hunk 5]
```diff
@@ -91,6 +98,7 @@ contract PollFactory is
         AccQueue messageAq = messageAqFactory.deploy(
             _treeDepths.messageTreeSubDepth
         );
+        messageAq.enqueue(NOTHING_UP_MY_SLEEVE);
```
Comment (Mikerah) [inline contracts/contracts/Poll.sol:101]: LGTM

File: contracts/contracts/TopupCredit.sol [hunk 6]
```diff
@@ -5,8 +5,8 @@ import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 
 contract TopupCredit is ERC20, Ownable {
-    uint8 private _decimals = 1;
-    uint256 MAXIMUM_AIRDROP_AMOUNT = 100000 * 10**_decimals;
+    uint8 private constant _decimals = 1;
```
Comment (Mikerah) [inline contracts/contracts/TopupCredit.sol:8]: LGTM

File: contracts/contracts/TopupCredit.sol [hunk 7]
```diff
@@ -5,8 +5,8 @@ import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
...
+    uint256 public constant MAXIMUM_AIRDROP_AMOUNT = 100000 * 10**_decimals;
```
Comment (Mikerah) [inline contracts/contracts/TopupCredit.sol:9]: LGTM

File: contracts/contracts/MACI.sol [hunk 8]
```diff
@@ -24,7 +24,7 @@ import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
 contract MACI is IMACI, DomainObjs, Params, SnarkCommon, Ownable {
     // The state tree depth is fixed. As such it should be as large as feasible
     // so that there can be as many users as possible.  i.e. 5 ** 10 = 9765625
-    uint8 public override stateTreeDepth = 10;
+    uint8 public constant override stateTreeDepth = 10;
```
Comment (Mikerah) [inline contracts/contracts/MACI.sol:27]: LGTM

**Review by Mikerah — CHANGES_REQUESTED** [review]

File: contracts/contracts/MACI.sol (hunk 2 above)
Comment (chaosma) [reply contracts/contracts/MACI.sol:226]: I think the limit for the voice credit is: `voiceCredit<sqrt(field_size)=147946756881789319005730692170996259609`
And this is larger than most (probably all) of the total hard cap of ERC20 token. (e.g. 1B tokens cap with 18 decimal)

BTW, what is the argument here?

Comment (Mikerah) [reply contracts/contracts/MACI.sol:226]: The argument is that you can either enforce this directly within the contract or remove the require statement and enforce it in the front-end and add documentation for developers who will rely on these contracts. The former is a more secure but the later saves a lot of gas. In either choice, the suggestion was to be more consistent with this and to properly document it.

**Comment by gurrpi** [comment]
There are two test failures on `contracts` package:

```
FAIL ts/__tests__/MACI.test.ts (18.714 s)
[531](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:532)
  MACI
[532](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:533)
    Deployment
[533](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:534)
      ✓ MACI.stateTreeDepth should be correct (9 ms)
[534](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:535)
    Signups
[535](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:536)
      ✓ should sign up users (1121 ms)
[536](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:537)
      ✓ signUp() shold fail when given an invalid pubkey (57 ms)
[537](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:538)
    Merging sign-ups should fail because of onlyPoll
[538](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:539)
      ✓ coordinator should not be able to merge the signUp AccQueue (20 ms)
[539](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:540)
    Deploy a Poll
[540](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:541)
      ✓ should set VKs and deploy a poll (498 ms)
[541](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:542)
      ✓ should set correct storage values (60 ms)
[542](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:543)
    Publish messages (vote + key-change)
[543](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:544)
      ✓ should publish a message to the Poll contract (3943 ms)
[544](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:545)
      ✓ shold not publish a message after the voting period (207 ms)
[545](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:546)
    Merge messages
[546](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:547)
      ✓ should revert if subtrees are not merged for StateAq (34 ms)
[547](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:548)
      ✓ coordinator should be able to merge the message AccQueue (1736 ms)
[548](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:549)
      ✕ the message root must be correct (28 ms)
✕ Should regenerate MaciState from on-chain information (422 ms)
[564](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:565)

● MACI › Merge messages › the message root must be correct
[566](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:567)

[567](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:568)
    expect(received).toEqual(expected) // deep equality
[568](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:569)

[569](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:570)
    Expected: "445596478217518740685967008242093136372266936138613472907310225044936378830"
[570](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:571)
    Received: "20938071508123277910401358646136004119033819121099042653495944919628428120495"
[571](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:572)

[572](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:573)
      460 |             const onChainMessageRoot = await messageAqContract.getMainRoot(MESSAGE_TREE_DEPTH)
[573](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:574)
      461 |             expect(onChainMessageRoot.toString())
[574](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:575)
    > 462 |                 .toEqual(maciState.polls[pollId].messageAq.mainRoots[MESSAGE_TREE_DEPTH].toString())
[575](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:576)
          |                  ^
[576](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:577)
      463 |         })
[577](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:578)
      464 |     })
[578](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:579)
      465 |
[579](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:580)

[580](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:581)
      at ts/__tests__/MACI.test.ts:462:18
[581](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:582)
      at step (ts/__tests__/MACI.test.ts:33:23)
[582](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:583)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
[583](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:584)
      at fulfilled (ts/__tests__/MACI.test.ts:5:58)
[584](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:585)

[585](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:586)
  ● MACI › Generate MaciState from contract › Should regenerate MaciState from on-chain information
[586](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:587)

[587](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:588)
    assert(received)
[588](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:589)

[589](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:590)
    Expected value to be equal to:
[590](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:591)
      true
[591](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:592)
    Received:
[592](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:593)
      false
[593](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:594)

[594](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:595)
      437 |             )
[595](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:596)
      438 |             const poll = maciState.polls[pollId]
[596](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:597)
    > 439 |             assert(
[597](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:598)
          |             ^
[598](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:599)
      440 |                 poll.messageAq.mainRoots[treeDepths.messageTreeDepth] ===
[599](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:600)
      441 |                 action.data.messageRoot
[600](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:601)
      442 |             )
[601](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:602)

[602](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:603)
      at ts/genMaciState.ts:439:13
[603](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:604)
      at step (ts/genMaciState.ts:44:23)
[604](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:605)
      at Object.next (ts/genMaciState.ts:25:53)
[605](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:606)
      at fulfilled (ts/genMaciState.ts:16:58)
[606](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:607)
          at runMicrotasks (<anonymous>)
[607](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:608)
          at runMicrotasks (<anonymous>)
[608](https://github.com/privacy-scaling-explorations/maci/actions/runs/3191117799/jobs/5650160646#step:7:609)
```

**Comment by ctrlc03** [comment]
Closing as changes were merged here https://github.com/privacy-scaling-explorations/maci/commit/f2cef9630709113302ea9a1855120094ca557472

**Pull request closed without merging**